capstone/src/test/java/com/google/step/servlets/RestaurantServletTest.java

```diff
@@ -0,0 +1,187 @@
+        helper.setEnvAuthDomain("google");
+
+        // Create a client to run the servlet
+        ServletRunner sr = new ServletRunner();
+        sr.registerServlet("restaurant", RestaurantServlet.class.getName());
+        ServletUnitClient sc = sr.newClient();
+
+        // Set params for the post request
+        WebRequest request = new PostMethodWebRequest("http://localhost:8080/restaurant");
+        request.setParameter("name", name);
+        request.setParameter("cuisine", cuisineList);
+        request.setParameter("story", story);
+        request.setParameter("phone", phone);
+        request.setParameter("website", website);
+        sc.getResponse(request);
+
+        // Simple query of the mock Datastore to check if the restaurant was entered correctly
+        Query query = new Query("RestaurantInfo")
+                .setFilter(new Query.FilterPredicate(
+                        "name", Query.FilterOperator.EQUAL, name));
+        PreparedQuery results = datastoreService.prepare(query);
+        Entity resultEntity = results.asSingleEntity();
+
+        assertEquals(name, resultEntity.getProperty("name"));
+        assertEquals("[Steakhouse,  American]", resultEntity.getProperty("cuisine").toString());
+        assertEquals(story, resultEntity.getProperty("story"));
+        assertEquals(phone, resultEntity.getProperty("phone"));
+        assertEquals(website, resultEntity.getProperty("website"));
+    }
+
+    /**
+     * Test the doPost method of RestaurantServlet with a user who is not logged in
+     * to ensure it will throw an exception.
+     * @throws IOException
+     * @throws SAXException if the ServletRunner parser cannot handle the request
+     */
+    @Test(expected = HttpException.class)
+    public void testDoPostLoginFailure() throws IOException, SAXException {
+        String name =  "Wildfire";
+        String story = "Swanky American chain serving steak, chops & seafood, plus burgers, sides & cocktails.";
+        String cuisineList = "Steakhouse, American";
+        String phone = "8472797900";
+        String website = "https://www.wildfirerestaurant.com";
+
+        // User is not logged in
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(false);
+
+        // Create a client to run the servlet
+        ServletRunner sr = new ServletRunner();
+        sr.registerServlet("restaurant", RestaurantServlet.class.getName());
+        ServletUnitClient sc = sr.newClient();
+
+        // Build request as normal
+        WebRequest request = new PostMethodWebRequest("http://localhost:8080/restaurant");
+        request.setParameter("name", name);
+        request.setParameter("cuisine", cuisineList);
+        request.setParameter("story", story);
+        request.setParameter("phone", phone);
+        request.setParameter("website", website);
+
+        // This getResponse should throw an exception because the user is not logged in.
+        sc.getResponse(request);
```

Comment:
This is more a question than a comment: does setting the status code to an error code cause an exception to be thrown? I'm just trying to understand what is happening here a bit more.

Reply:
Good catch, I looked it up and the setStatus function that will get called if the user isn't logged in should just set the status, not throw an exception. I think the exception is thrown by the test as it runs. I updated the documentation.